cloudinit/reporting/handlers.py

```diff
@@ -81,34 +83,79 @@ def __init__(
         super(WebHookHandler, self).__init__()
 
         if any([consumer_key, token_key, token_secret, consumer_secret]):
-            self.oauth_helper = url_helper.OauthUrlHelper(
+            oauth_helper = url_helper.OauthUrlHelper(
                 consumer_key=consumer_key,
                 token_key=token_key,
                 token_secret=token_secret,
                 consumer_secret=consumer_secret,
             )
+            self.readurl = oauth_helper.readurl
         else:
-            self.oauth_helper = None
+            self.readurl = url_helper.readurl
         self.endpoint = endpoint
         self.timeout = timeout
         self.retries = retries
         self.ssl_details = util.fetch_ssl_details()
 
+        self.flush_requested = Event()
+        self.queue = queue.Queue()
+        self.event_processor = threading.Thread(target=self.process_requests)
+        self.event_processor.daemon = True
+        self.event_processor.start()
+
+    def process_requests(self):
+        consecutive_failed = 0
+        while True:
+            if self.flush_requested.is_set() and consecutive_failed > 2:
+                # At this point the main thread is waiting for the queue to
+                # drain. If we have a queue of events piled up and recent
+                # events have failed, lets not waste time trying to post
+                # the rest, especially since a long timeout could block
+                # cloud-init for quite a long time.
+                LOG.warning(
+                    "Multiple consecutive failures in WebHookHandler. "
+                    "Cancelling all queued events."
+                )
+                while not self.queue.empty():
+                    self.queue.get_nowait()
+                    self.queue.task_done()
+                consecutive_failed = 0
+            args = self.queue.get(block=True)
```

Comment:
Won't this blocking `queue.get()` cause a deadlock when `.join()` is hit? I tried to put together a simple example to demonstrate:

```python3
from threading import Thread, Event
from queue import Queue

e = Event()
q = Queue()

def getter():
    print("entering getter")
    while True:
        print("before blocking get")
        val = q.get(block=True)
        print(f"after blocking get, received {val}")


Thread(target=getter).start()

e.set()
q.put("one")
q.join()
e.clear()
```
example:
```bash
arc~ python3 queue-model.py 
entering getter
before blocking get
after blocking get, received one
before blocking get
```

One way around this (I'm sure there are many) would be to put a sentinal value in the queue that, when received, causes the thread to exit.

It's entirely possible I oversimplified my example too much and missed something. Let me know.

Reply:
Yeah, `Queue.join()` is a little weird. According to the docs:
> If a [join()](https://docs.python.org/3/library/queue.html#queue.Queue.join) is currently blocking, it will resume when all items have been processed (meaning that a [task_done()](https://docs.python.org/3/library/queue.html#queue.Queue.task_done) call was received for every item that had been [put()](https://docs.python.org/3/library/queue.html#queue.Queue.put) into the queue).

So if all the items in the queue have been processed, join() should unblock. I think I have that covered in the `self.queue.task_done()` calls on lines 141 and 121.

To update your example, I added a `.task_done()` after the print in the while loop and made the thread a daemon, which I think more accurately reflects the code. With these changes, I don't see any deadlock.

```python
from queue import Queue
from threading import Event, Thread

e = Event()
q = Queue()


def getter():
    print("entering getter")
    while True:
        print("before blocking get")
        val = q.get(block=True)
        print(f"after blocking get, received {val}")
        q.task_done()


t = Thread(target=getter)
t.daemon = True
t.start()

e.set()
q.put("one")
print("before join")
q.join()
print("after join")
e.clear()
```

I welcome the deeper look though, and let me know if there are any cases I missed.

Reply:
> To update your example, I added a `.task_done()` after the print in the while loop and made the thread a daemon, which I think more accurately reflects the code. With these changes, I don't see any deadlock.

Ah yes, you're right, a manual exit from the thread is unnecessary. The `daemon=True` setting is what I missed in my example. (I had `.task_done()` locally, but accidentally deleted that while cleaning up to share.)